### PR TITLE
fqdn: don't allocate a new DNS request ID

### PR DESCRIPTION
**Pull request by bimmlerd**
The DNS proxy allocated a random request ID before forwarding the query to the upstream server and restored the original request ID before sending the response back to the client. Unfortunately, since the request ID was chosen uniformly at random, there was no guarantee that there would be no repetition. Since request IDs are uint16 and due to the birthday paradox, the chance of repetition is unintuitively high. Indeed, for a window of 200 queries, the chance of seeing a repeated ID is one in four.

Luckily, we don't depend on ID uniqueness in general - since the DNS proxy serves many clients concurrently, it would be unrealistic to assume that they'd somehow managed to avoid ID collisions. However, our shared client infrastructure currently _does_ assume that for a five-tuple, the request IDs do not overlap. A single client sending multiple differing queries with the same ID is less likely, since it typically uses these IDs to efficiently match responses to queries. (Retries are possible, but not the focus of this change).

Now, given the above it's suprising to see that, since 804d5f0f22 (fqdn: Fix goroutine leak in transparent-mode) we regularly observe DNS errors around duplicate request IDs. Often enough, that 3bdfd9dc4f (dnsproxy: Pick up cilium/dns with ID retry logic) added retries to allocate another random ID if we had already hit this one.

The root cause, however, is that due to dual-stack networking and multiple entries in the DNS search path, a single application-level name lookup can be translated into many (say, 10) actual DNS queries on the wire. Dual stack contributes with a factor 2 (A/AAAA queries). Even worse, each search path entry is checked (depending on the DNS lib being used), hence the the number of queries is multiplied by the number of entries in the search path.

Since all of these DNS messages are being sent concurrently (as a latency optimization) and from the same five-tuple, the shared client would every now and then see ID duplication due to the random allocation by the proxy. The probability of seing a duplicate is not high (for 10 queries, it's merely one in ~1500), but that's enough to see occasional drops due to this duplication.

Long story short, let's not randomly allocate new DNS request IDs, as it does more harm than good. For a given five-tuple, we don't expect to see duplicates, hence there's also no reason to allocate.

Note that the matter of retries is also resolved, thanks to the previously introduced patch of allocating a fresh ID when a duplicate does occur.

```release-note
Fix a low-probability issue where the DNS proxy could occasionally drop DNS queries due to "duplicate request id" errors.
```

### CODE REVIEW

**Review by gandro — APPROVED** [review]
Thanks for the detailed description! While reading, I wondered why the ID is randomized in the first place. It seems to come from the original PR adding the feature (https://github.com/cilium/cilium/pull/5641, in particular the source code comment highlighted by the comment here https://github.com/cilium/cilium/pull/5641/files#r233983909) and seems to exist because in non-transparent mode there used to be only one single shared upstream DNS client for all downstream DNS clients.

For non-transparent mode, having only a single shared DNS client for all downstreams without some kind of ID randomization would probably cause more harm than randomized IDs. However, it seems that since https://github.com/cilium/cilium/pull/22619 we stopped having a single shared DNS client in non-transparent mode, and instead create a new client for each request. This makes ID randomization pretty pointless in the non-transparent mode, since there shouldn't be any collisions if there are is no shared client.

In transparent mode you already argued why it is probably better to to keep the original request ID. So transparent mode (where there is a bit of client sharing still) should be fine to.

Which is a long way of saying: Yes, I concur that we no longer need ID randomization!

**Comment by espringsteen** [comment]
Is this something that should be backported to 1.15 as well?

**Comment by bimmlerd** [comment]
> Is this something that should be backported to 1.15 as well?

It could, but we don't usually backport non-major bugfixes like this one further than one branch. IMO, it doesn't match the necessary criteria: https://docs.cilium.io/en/latest/contributing/release/backports/#backport-criteria-for-x-y-1-z-and-x-y-2-z
